File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.*;
import java.lang.Math.*;

class Solution {
    String leftFinger = "*";
    String rightFinger = "#";
    
    public String solution(int[] numbers, String hand) {
        String answer = "";
        
        List<String> leftLow = List.of("1","4","7");
        List<String> rightLow = List.of("3","6","9");
        
        for (int i=0;i<numbers.length;i++) {
            String number = Integer.toString(numbers[i]);
            if (leftLow.contains(number)) {
                leftFinger = number;
                answer += "L";
            } else if (rightLow.contains(number)) {
                rightFinger = number;
                answer += "R";
            } else {
                answer += getMidFinger(number, hand);
            }
        }
        return answer;
    }
    
    private String getMidFinger( 
        String number,
        String hand
    ) {
        String[][] keypad = {{"1","2","3"},
                             {"4","5","6"},
                             {"7","8","9"},
                             {"*","0","#"}};
        
        int[] numberPos = new int[2];
        int[] leftPos = new int[2];
        int[] rightPos = new int[2];
        
        for (int row=0;row<keypad.length;row++){
            for (int col=0;col<keypad[row].length;col++){
                String word = keypad[row][col];
                if (word.equals(number)) {
                    numberPos[0] = row;
                    numberPos[1] = col;
                }
                if (word.equals(leftFinger)) {
                    leftPos[0] = row;
                    leftPos[1] = col;
                }
                if (word.equals(rightFinger)) {
                    rightPos[0] = row;
                    rightPos[1] = col;
                }
            }
        }
        
        int leftDistance = Math.abs(numberPos[0]-leftPos[0])
            + Math.abs(numberPos[1]-leftPos[1]);
        
        int rightDistance = Math.abs(numberPos[0]-rightPos[0]) 
            + Math.abs(numberPos[1]-rightPos[1]);
        
        if (leftDistance == rightDistance) {
            if (hand.equals("right")) {
                rightFinger = number;
                return "R";
            } else {
                leftFinger = number;
                return "L";
            }
        } else if (leftDistance > rightDistance) {
            rightFinger = number;
            return "R";
        } else {
            leftFinger = number;
            return "L";
        }
        
    }
}
                
                

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    pad = [[1,2,3],[4,5,6],[7,8,9],['*',0,'#']]
    loc = {pad[r][c]:[r, c] for r in range(len(pad)) for c in range(len(pad[r]))}
    h = {'L':'*', 'R':'#'}
    for i in range(len(numbers)):
        if numbers[i] in [1,4,7]:
            answer += 'L'
            h['L'] = numbers[i]
        elif numbers[i] in [3,6,9]:
            answer += 'R'
            h['R'] = numbers[i]
        # 2580이면 직전 위치로부터 맨해튼 거리 구하기
        else:
            l_dis = abs(loc[h['L']][0]-loc[numbers[i]][0])+abs(loc[h['L']][1]-loc[numbers[i]][1])
            r_dis = abs(loc[h['R']][0]-loc[numbers[i]][0])+abs(loc[h['R']][1]-loc[numbers[i]][1])
            print(h, l_dis, r_dis)
            if l_dis == r_dis:
                x = hand[0].upper()
                answer += x
                h[x] = numbers[i]
            elif l_dis > r_dis:
                answer += 'R'
                h['R'] = numbers[i]
            else:
                answer += 'L'
                h['L'] = numbers[i]
            
            
    
    return answer